crates/subspace-block-relay/src/protocol/full_block.rs

```diff
@@ -0,0 +1,512 @@
+    fn validate(
+        &self,
+        _context: &mut dyn ValidatorContext<Block>,
+        sender: &PeerId,
+        mut data: &[u8],
+    ) -> ValidationResult<Block::Hash> {
+        // TODO: substrate requires decoding twice, once during validate and again
+        // during the processing.
+        if let Err(err) = BlockAnnouncement::<Block>::decode(&mut data) {
+            warn!(
+                target: LOG_TARGET,
+                "FullBlockRelayValidator::validate(): peer = {:?}, decode failed: {:?}",
+                sender,
+                err
+            );
+            ValidationResult::Discard
+        } else {
+            ValidationResult::ProcessAndKeep(topic::<Block>())
+        }
+    }
+
+    fn message_expired<'a>(&'a self) -> Box<dyn FnMut(Block::Hash, &[u8]) -> bool + 'a> {
+        Box::new(move |topic, data| {
+            trace!(
+                target: LOG_TARGET,
+                "FullBlockRelayValidator::message_expired(): topic = {:?}, data = {:?}",
+                topic,
+                data
+            );
+            !self.pending_announcements.lock().contains(data)
+        })
+    }
+
+    fn message_allowed<'a>(
+        &'a self,
+    ) -> Box<dyn FnMut(&PeerId, MessageIntent, &Block::Hash, &[u8]) -> bool + 'a> {
+        Box::new(move |peer, intent, topic, data| {
+            let i = match intent {
+                MessageIntent::Broadcast => "Broadcast",
+                MessageIntent::ForcedBroadcast => "ForcedBroadcast",
+                MessageIntent::PeriodicRebroadcast => "PeriodicRebroadcast",
+            };
+            trace!(
+                target: LOG_TARGET,
+                "FullBlockRelayValidator::message_allowed(): topic = {:?}, data = {:?}, \
+                peer = {:?}, intent = {:?}",
+                topic,
+                data,
+                peer,
+                i
+            );
+
+            let mut pending_announcements = self.pending_announcements.lock();
+            pending_announcements.remove(data)
+        })
+    }
+}
+
+/// Block relay message topic.
+fn topic<Block: BlockT>() -> Block::Hash {
+    <<Block::Header as HeaderT>::Hashing as HashT>::hash(ANNOUNCE_PROTOCOL.as_bytes())
+}
+
+/// Initializes the full block relay specific config.
+pub(crate) fn init_full_block_config(config: &mut Configuration) -> Receiver<IncomingRequest> {
+    let mut cfg = NonDefaultSetConfig::new(ANNOUNCE_PROTOCOL.into(), 1024 * 1024);
+    cfg.allow_non_reserved(25, 25);
+    config.network.extra_sets.push(cfg);
+
+    let (request_sender, request_receiver) = mpsc::channel(INBOUND_QUEUE_SIZE);
+    config
+        .network
+        .request_response_protocols
+        .push(RequestResponseConfig {
+            name: SYNC_PROTOCOL.into(),
+            fallback_names: Vec::new(),
+            max_request_size: 1024 * 1024,
+            max_response_size: 16 * 1024 * 1024,
+            request_timeout: Duration::from_secs(60),
```

Comment:
All of these are likely way too large. Both request/response (even for sync they are smaller) and timeout of 60 seconds is too much.